audio-router/src/main/java/com/twilio/audio_router/AudioDeviceSelector.java

```diff
@@ -0,0 +1,416 @@
+        return selectedDevice != null
+                ? new AudioDevice(selectedDevice.type, selectedDevice.name)
+                : null;
+    }
+
+    /**
+     * Retrieves the current list of available {@link AudioDevice}s.
+     *
+     * @return the current list of {@link AudioDevice}s
+     */
+    public @NonNull List<AudioDevice> getAudioDevices() {
+        ThreadUtils.checkIsOnMainThread();
+        return Collections.unmodifiableList(new ArrayList<>(availableAudioDevices));
+    }
+
+    private void setAudioFocus() {
+        // Request audio focus before making any device switch.
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            AudioAttributes playbackAttributes =
+                    new AudioAttributes.Builder()
+                            .setUsage(AudioAttributes.USAGE_VOICE_COMMUNICATION)
+                            .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
+                            .build();
+            AudioFocusRequest focusRequest =
+                    new AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN_TRANSIENT)
+                            .setAudioAttributes(playbackAttributes)
+                            .setAcceptsDelayedFocusGain(true)
+                            .setOnAudioFocusChangeListener(i -> {})
+                            .build();
+            audioManager.requestAudioFocus(focusRequest);
+        } else {
+            audioManager.requestAudioFocus(
+                    focusChange -> {},
+                    AudioManager.STREAM_VOICE_CALL,
+                    AudioManager.AUDIOFOCUS_GAIN_TRANSIENT);
+        }
+        /*
+         * Start by setting MODE_IN_COMMUNICATION as default audio mode. It is
+         * required to be in this mode when playout and/or recording starts for
+         * best possible VoIP performance. Some devices have difficulties with speaker mode
+         * if this is not set.
+         */
+        audioManager.setMode(AudioManager.MODE_IN_COMMUNICATION);
+    }
+
+    private class WiredHeadsetReceiver extends BroadcastReceiver {
+        private static final int STATE_UNPLUGGED = 0;
+        private static final int STATE_PLUGGED = 1;
+
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            ThreadUtils.checkIsOnMainThread();
+            int state = intent.getIntExtra("state", STATE_UNPLUGGED);
+            if (state == STATE_PLUGGED) {
+                wiredHeadsetAvailable = true;
+                Log.d(TAG, "Wired Headset available");
+                if (AudioDeviceSelector.this.state == State.ACTIVE) {
+                    userSelectedDevice = WIRED_HEADSET_AUDIO_DEVICE;
+                }
+                enumerateDevices();
+            } else {
+                wiredHeadsetAvailable = false;
+                enumerateDevices();
+            }
+        }
+    }
+
+    private void enumerateDevices() {
+        availableAudioDevices.clear();
```

Comment:
Let's add a thread checker here to make sure that this is on the main thread. I want to ensure that we meet the contract I stated above when calling the listener.